Mvs 267 create review and submit registration (household) page as a vue component

dmcook5208 opened the pull request:
## :flipper: [JIRA Sprint Card]
https://mass-immunization-system.atlassian.net/browse/MVS-267

## :newspaper: [Work Description]
Updated the HouseholdReviewSubmit page to display the registration information for the household (common address, contact info), including the personal information for up to 20 patients.

## :vertical_traffic_light: [Testing/QA Notes]
- Proceed down the "Register my household" workflow, selecting various numbers for the patients within the household (2-20).
- Proceed to the "Review and submit" page
- Verify that the number of patients presented (personal info sections) is consistent with the number of patients you specified in the first step of the workflow
NOTE:  This page only displays the actual entered data for household member #1.  I am creating a new card (MVS-385) to handle the entry of household members #2-20, at which point, I will make this data presentable in the "review and submit" page, as well.

david-fraley commented:
I'm assuming in your QA notes, the separate card you mention is for patients 2-20 and not 3-20... Please adjust!

david-fraley commented on PatientRegistrationApp/src/App.vue:
```diff
@@ -331,7 +331,9 @@ export default {
-			this.goToPage(config.registrationPages.HOUSEHOLD_PERSONAL_INFO_1_PAGE);
+			this.$refs.householdcontactinfo.verifyFormContents() ?
+			this.goToPage(config.registrationPages.HOUSEHOLD_PERSONAL_INFO_1_PAGE) :
+			this.goToPage(config.registrationPages.HOUSEHOLD_CONTACT_INFO_PAGE);
```
What does this variable name actually refer to -- `HOUSEHOLD_PERSONAL_INFO_1_PAGE`. At any rate, it should be renamed to be more clear and specific for what it is.

david-fraley replied:
Upon reviewing the config file, is `HOUSEHOLD_REGISTER_NUMBER_PAGE` referencing phone number? That should be explicit then...

dmcook5208 replied:
HOUSEHOLD_PERSONAL_INFO_1_PAGE is the identifier for the HouseholdPersonalInfo_1.vue page.
HOUSEHOLD_REGISTER_NUMBER_PAGE is the identifier for the HouseholdRegisterNumber.vue page.

If/when we create design specifications, we could clearly associate the vue component names (and enumerations) to the layout of the page so that there is no confusion.

david-fraley replied:
I'm still not following the `_1.vue` page syntax. Do you mean that this is the personal info for the first person? I think the name we're using could be more descriptive.

dmcook5208 replied:
I changed the enumeration to HOUSEHOLD_PERSONAL_INFO_PATIENT_1_PAGE; this is the page that allows you to enter the personal information of patient #1.

david-fraley replied:
More clear, thanks

david-fraley commented on PatientRegistrationApp/src/components/HouseholdContactInfo.vue:
```diff
@@ -1,133 +1,146 @@
+			var valid = true
+			var message = "Woops! You need to enter the following fields:"
 		
+			if(this.primaryPhoneNumber == "") {
```
For all of these fields, I think we should also check if the prop is null as well as empty (you're doing empty rn). I'm pretty sure Vue handles these differently since most javascript-y languages treat null/empty differently. 

That might take the form as `if(!this.primaryPhoneNumber && this.primaryPhoneNumber == "") {}`. I was reading online and this seems like a cleaner implementation - the example explicitly says `this.$options.propData` so I figure `options` or `propData` would be replaced with `primaryPhoneNumber`. 

This also comes with another question - when you make these fields, do you initialize or default the prop to be null? The bottom of this page indicates possibly yes, but I'm not sure if they're using an eventbus https://forum.vuejs.org/t/is-there-any-way-to-check-if-all-props-passed-the-validation/26547/2

dmcook5208 replied:
This wasn't part of my change - if GitHub says it is, perhaps it was due to a manual merge or rearrangement of lines.  Create a separate card for this?

david-fraley replied:
Yea, it's probably because of how you wrote some code and shifted something down. A separate card would work!

dmcook5208 replied:
Created MVS-396

david-fraley commented on PatientRegistrationApp/src/components/HouseholdContactInfo.vue:
```diff
@@ -1,133 +1,146 @@
+			emailRules: [
+				(v) =>
+				/^[\s]*$|.+@.+\..+/.test(v) ||
+				"E-mail must be in the format example@example.com",
```
We should definitely change the .com to be .<text>. My personal email ends in .us, non-profits are .org, and educational institutions are .edu so this rule becomes quite restrictive

dmcook5208 replied:
This wasn't part of my change - if GitHub says it is, perhaps it was due to a manual merge or rearrangement of lines.  Create a separate card for this?

david-fraley replied:
A separate card works!

dmcook5208 replied:
Actually, the code allows any extension beyond the "." (so .us, .edu, .gov, all work).  Do we just need to update the user feedback that is displayed?  
I created MVS-395 to decide this.

david-fraley replied:
Oh, if that's the case then yes only the user feedback should be clarified. I'm assuming we're also fine if the email is, for example, david.fraley@something.com right? I added that into the card.

The comment could read: `"E-mail must be in the format example@example.com, .org, .edu, .us, etc."` Is that weird - I'll leave final wording up to you

david-fraley commented on PatientRegistrationApp/src/components/HouseholdContactInfo.vue:
```diff
@@ -1,133 +1,146 @@
+				/^[\s]*$|.+@.+\..+/.test(v) ||
+				"E-mail must be in the format example@example.com",
+			],
+			primaryPhoneNumber: '',
```
Do we have phone-number rules?

dmcook5208 replied:
MVS-257 and MVS-259 are for phone number validation (to do)

david-fraley replied:
Awesome thanks

david-fraley commented on PatientRegistrationApp/src/components/HouseholdContactInfo.vue:
```diff
@@ -1,133 +1,146 @@
+				message += "*Follow up consent"
+				valid = false
+			}
+			if (valid == false) {
```
Would it be possible to nest this conditional and the one above? I'm not positive when they get called so I'm not sure I can provide more guidance on that, without context

dmcook5208 replied:
This wasn't part of my change - if GitHub says it is, perhaps it was due to a manual merge or rearrangement of lines.  Create a separate card for this?

However, I think the logic is fine - if any one of the conditions above is false, the overall validation of the page is false (but we want to know exactly which conditions fail so that we can tell the user what to correct)

david-fraley commented on PatientRegistrationApp/src/components/HouseholdEmergencyContact.vue:
```diff
@@ -63,43 +62,42 @@ export default {
       }
-      EventBus.$emit('DATA_EMERGENCY_CONTACT_INFO_PUBLISHED', emergencyContactPayload)
-    },*/
+      EventBus.$emit('DATA_HOUSEHOLD_EMERGENCY_CONTACT_INFO_PUBLISHED', householdEmergencyContactPayload)
```
Based on the name, this looks like a good example of application state!

dmcook5208 replied:
It's just a message to send household emergency contact data to the "review and submit" page.  If it's confusing, I can change PUBLISHED to ENTERED.  Until that message is able to be sent (required data has been entered), it is not possible to proceed in the workflow...so yes, it is technically a state, but I don't see a reason to add a separate application state variable to manage it.

david-fraley commented on PatientRegistrationApp/src/components/HouseholdEmergencyContact.vue:
```diff
@@ -63,43 +62,42 @@ export default {
+      var valid = true
+      var message = "Woops! You need to enter the following fields:"
+      
+      if(this.householdEmergencyContactFamilyName == "")  {
```
Same comments about null and empty check as above

dmcook5208 replied:
This wasn't part of my change - if GitHub says it is, perhaps it was due to a manual merge or rearrangement of lines.  Create a separate card for this?

dmcook5208 replied:
Created MVS-396

david-fraley commented on PatientRegistrationApp/src/components/HouseholdEmergencyContact.vue:
```diff
@@ -63,43 +62,42 @@ export default {
+        valid = false
+      }
+      
+      if (this.householdEmergencyContactPhoneNumber == "") {
```
Same as above for nesting this and the `valid==false` conditional

dmcook5208 replied:
This wasn't part of my change - if GitHub says it is, perhaps it was due to a manual merge or rearrangement of lines. Create a separate card for this?

However, I think the logic is fine - if any one of the conditions above is false, the overall validation of the page is false (but we want to know exactly which conditions fail so that we can tell the user what to correct)

david-fraley replied:
I think a separate card would suffice. Probably doing too many things on a given branch/card that are out of scope. 

If you think the logic is good, good with me. Just wanted to check!

david-fraley commented on PatientRegistrationApp/src/components/HouseholdRegisterNumber.vue:
```diff
@@ -34,7 +34,7 @@ watch: {
 				householdCount: this.e1
 			}
-			EventBus.$emit('DATA_HOUSHOLD_COUNT_UPDATED', householdCountPayload)
+			EventBus.$emit('DATA_HOUSEHOLD_COUNT_UPDATED', householdCountPayload)
```
Another application state example!

dmcook5208 replied:
It's just a message to send the "number of household members" data to the App.vue page when a value changes.

david-fraley replied:
Yep just pointing out like earlier!

david-fraley commented on PatientRegistrationApp/src/components/HouseholdReviewSubmit.vue:
```diff
@@ -1,22 +1,162 @@
+				<v-card class="d-flex flex-wrap" flat>
+					<v-card class="pa-2" flat min-width=33%>
+						<div class="font-weight-medium primary--text">Personal Info: Household Member #1</div>
+						<div class="font-weight-medium">Name:  <span class="font-weight-regular">{{dataHouseholdPersonalInfo[0].householdFamilyName}}, 
```
Is the `...[0]...` basically why your code only shows the first patient? If so, are you trying to set up a framework to accommodate the future card or is all of that work also handled in this new card?

david-fraley replied:
Some type of loop might be good for it.?

dmcook5208 replied:
I do patient 1 separately (because they have a different emergency contact), but I do a for loop for patients 2-20 (starting at line 56).  MVS-385 will take care of populating the data for patients 2-20.

dmcook5208 commented:
> I'm assuming in your QA notes, the separate card you mention is for patients 2-20 and not 3-20... Please adjust!

Whoops!  I fixed that and added a reference to the separate card (MVS-385)

david-fraley commented on PatientRegistrationApp/src/components/HouseholdReviewSubmit.vue:
```diff
@@ -1,22 +1,162 @@
+						:key="n"
+						class="pa-2" flat min-width=33%>
+						<div class="font-weight-medium primary--text">Personal Info: Household Member #{{n+1}}</div>
+						<div class="font-weight-medium">Name:  <span class="font-weight-regular"><!--{{dataHouseholdPersonalInfo[n].householdFamilyName}}, 
```
Can you clarify the commented out code - this is what you meant which will be populated in MVS-385?

dmcook5208 replied:
Yes, exactly

david-fraley reviewed (changes requested):
Mostly questions and card creation!

david-fraley reviewed (approved):
Please fix the User Feedback for the email address stuff before merging!
